src/tables/service_periods.py

```diff
@@ -20,3 +22,82 @@ def __init__(self, user=None, passwd=None, hostname="localhost", db_name="apertu
                 ternary SMALLINT NOT NULL CHECK ( (ternary <= 3) AND (ternary >= 1) ),
                 UNIQUE (month, year, ternary)
             );"""])
+
+    
+    def write_table(self, dates):
+        # Dates: list of datetime dates. Only important value in the dates are
+        # month and year.
+        data = []
+        for date in dates:
+            data.append([date.month, date.year, self._get_ternary(date.month)])
+        df = pandas.DataFrame(data, columns=self._expected_cols)
+        # service_key is BIGSERIAL so there should never be a conflict.
+        return self._write_table(df)  
+
+
+    def query_month_year(self, month, year):
+        # Find a service_periods row that matches the month and year.
+        # Note that ternary isn't included because that's derived from month.
+
+        if self._engine is None:
+            self._print("ERROR: No sql engine.")
+            return None
+        service_key = None
+
+        sql = "".join(["SELECT * FROM ", self._schema, ".", self._table_name,
+                       " WHERE month=", str(month), " AND year=", str(year), ";"
+                       ])
+        try:
+            with self._engine.connect() as con:
+                result = con.execute(sql)
+                if result.rowcount != 0:
+                    return result.first()['service_key']
+        except SQLAlchemyError as error:
+            print("SQLAlchemy: ", error)
+            return None
+
+
+    def insert_one(self, month, year):
+        # Insert a row and return the id.
+        # Honestly should be using write_table() but the RETURNING
+        # functionality isn't built into write_table() and I'm too lazy to
+        # add that functionality right now so I'll save it for a TODO.
+        if self._engine is None:
+            self._print("ERROR: No sql engine.")
+            return None
+
+        ternary = self._get_ternary(month)
+        sql = "".join(["INSERT INTO ", self._schema, ".", self._table_name,
+                       " (month, year, ternary) VALUES (",
+                       str(month), ", ", str(year), ", ", str(ternary), ")",
+                       " RETURNING service_key;"])
+        try:
+            with self._engine.connect() as con:
+                result = con.execute(sql)
+                return result.first()[0]
+        except SQLAlchemyError as error:
+            print("SQLAlchemy: ", error)
+            return None
+
+
+    def query_or_insert(self, month, year):
```

Comment:
Any suggestions? I think this is a fair method name, tho documentation would be good yes.

Reply:
Maybe just `get`? It would make it more obvious what it's doing to someone who doesn't know what it does, and if someone is curious about it, they can look into it more.

Reply:
`special_get` could be another option, it better reflects how this getter will also create if not existing, and it's main purpose of returning a value is more clear.

Reply:
This also inserts a new row if a row doesn't already exist, so `get` would be inappropriate.

Reply:
I agree, `get` is not great. Maybe we can wait till someone else takes a look at this and get a third set of eyes.

Reply:
I don't like `special_get`. If I see `special_get` in a PR I would probably make the same comment you are making lmao.

Reply:
imo query_or_insert is pretty explicit about what it does, it either queries or it inserts. I mean, a more descriptive name would be `query_and_insert_if_failed` but that's way too long so.

Reply:
My main issue with the name is that if you don't already know what it does, it's not very obvious what it does, since a query would be expected to return a value, but an insert is expected to return a boolean, and the `or` just makes it seem very unpredictable unless you pull up the docs, which I think is undesirable. The method does kinda do two things (and I do see the convenience of having it do them both at the same time), so the longer name seems fitting.

Reply:
I do agree with you that it's not a great name, but I can't think of a better one that is more descriptive and shorter than this one.

Reply:
`query_or_insert` is fine imo, it's clear what it does. Django uses similar conjoined method names like `get_or_create` for those cases.